firewood/src/merkle/node.rs

```diff
@@ -451,87 +444,83 @@ impl Storable for Node {
             }
     }
 
     fn serialize(&self, to: &mut [u8]) -> Result<(), ShaleError> {
         trace!("[{self:p}] Serializing node");
-        let mut cur = Cursor::new(to);
-
-        let mut attrs = match self.root_hash.get() {
-            Some(h) => {
-                cur.write_all(&h.0)?;
-                NodeAttributes::ROOT_HASH_VALID
-            }
-            None => {
-                cur.write_all(&[0; 32])?;
-                NodeAttributes::empty()
-            }
+        let mut cursor = Cursor::new(to);
+
+        let root_hash = self.root_hash.get().map(|hash| hash.0);
+        let encoded = self.encoded.get();
+        let encoded_len = encoded.map(Vec::len).map(|len| len as u64);
+
+        let mut attrs = match (encoded_len, self.is_encoded_longer_than_hash_len.get()) {
+            // the raw encoded-node is longer than a hash
+            (None, Some(true)) => NodeAttributes::LONG,
+            (Some(len), _) if len > TRIE_HASH_LEN as u64 => NodeAttributes::LONG,
+            //
+            (Some(_), _) | (None, Some(false)) => NodeAttributes::IS_ENCODED_BIG_VALID,
+            (None, None) => NodeAttributes::empty(),
         };
 
-        let mut encoded_len: u64 = 0;
-        let mut encoded = None;
-        if let Some(b) = self.encoded.get() {
-            attrs.insert(if b.len() > TRIE_HASH_LEN {
-                NodeAttributes::LONG
-            } else {
-                encoded_len = b.len() as u64;
-                encoded = Some(b);
-                NodeAttributes::IS_ENCODED_BIG_VALID
-            });
-        } else if let Some(b) = self.is_encoded_longer_than_hash_len.get() {
-            attrs.insert(if *b {
-                NodeAttributes::LONG
-            } else {
-                NodeAttributes::IS_ENCODED_BIG_VALID
-            });
+        if root_hash.is_some() {
+            attrs.insert(NodeAttributes::ROOT_HASH_VALID);
         }
 
-        #[allow(clippy::unwrap_used)]
-        cur.write_all(&[attrs.bits()]).unwrap();
+        let root_hash = root_hash.unwrap_or_default();
 
-        cur.write_all(&encoded_len.to_le_bytes())?;
-        if let Some(encoded) = encoded {
-            cur.write_all(encoded)?;
-            let remaining_len = ENCODED_MAX_LEN - encoded_len as usize;
-            cur.write_all(&vec![0; remaining_len])?;
-        } else {
-            cur.write_all(&[0; ENCODED_MAX_LEN])?;
-        }
+        let (encoded_len, encoded) = {
+            let encoded_len = encoded_len.filter(|len| *len <= TRIE_HASH_LEN as u64);
+            let encoded = std::array::from_fn({
+                let mut iter = encoded_len
+                    .and(encoded)
+                    .map(Vec::as_slice)
+                    .unwrap_or_default()
+                    .iter()
+                    .copied();
+
+                move |_| iter.next().unwrap_or_default()
+            });
+
+            (encoded_len.unwrap_or_default(), encoded)
+        };
```

Comment:
I scratched my head on this one for a while; I think this is equivalent code but my confidence is not 100%:
```suggestion
        // have: encoded_len: Option<u64> and encoded: Option<&Vec<u8>>
        // If either of these are None, encoded is [0u8; 32]
        // encoded_len just unwrapped
        let (encoded_len, encoded) = match (encoded_len, encoded) {
            (Some(encoded_len), Some(encoded)) => (encoded_len, encoded.as_slice().try_into().unwrap_or([0; TRIE_HASH_LEN])),
            _ => (0, [0u8; TRIE_HASH_LEN]),
        };
```